### ISSUE TITLE
Kudu Sink looking for Schema Registry even if not needed?

### ISSUE THREAD

**Issue by pmid** [Landoop/stream-reactor#421]
Hello,

I'm trying to run the kudu sink on a Cloudera cluster ( (Cloudera) kafka 0.11, no Schema Registry installed, kudu sink 1.0.0).

I setup a sample table, following the docs:

```
CREATE TABLE default.kudu_test (id INT,random_field STRING, PRIMARY KEY(id))
PARTITION BY HASH PARTITIONS 2
STORED AS KUDU
TBLPROPERTIES(
  'kudu.num_tablet_replicas' = '1' );

```

I'm not using Avro, and I didn't enable AUTOCREATE in KCQL.
The sink configuration is:

```
name=kudu-sink
connector.class=com.datamountaineer.streamreactor.connect.kudu.sink.KuduSinkConnector
tasks.max=1
connect.progress.enabled=true
topics=my_topic
connect.kudu.master=kudu_ip:7051
connect.kudu.kcql=INSERT INTO impala::default.kudu_test SELECT * FROM my_topic
```
  
When I start KafkaConnect (standalone mode) I get:
  
[2018-03-30 17:24:40,000] INFO Connecting to Kudu Master at <kudu_ip>:7051 (com.datamountaineer.streamreactor.connect.kudu.sink.KuduWriter$:43)
[2018-03-30 17:24:40,219] INFO Initialising Kudu writer (com.datamountaineer.streamreactor.connect.kudu.sink.KuduWriter:51)
[2018-03-30 17:24:40,503] INFO **Found the following tables in Kudu, impala::default.kudu_test, .....** (com.datamountaineer.streamreactor.connect.kudu.sink.DbHandler$:53)
[2018-03-30 17:24:40,513] ERROR Task kudu-sink-0 threw an uncaught and unrecoverable exception (org.apache.kafka.connect.runtime.WorkerTask:148)
**java.lang.NoClassDefFoundError: io/confluent/kafka/schemaregistry/client/rest/RestService**
        at com.datamountaineer.streamreactor.connect.schemas.SchemaRegistry$.getSubjects(SchemaRegistry.scala:61)
        at com.datamountaineer.streamreactor.connect.kudu.sink.DbHandler$.createTables(DbHandler.scala:96)
        at com.datamountaineer.streamreactor.connect.kudu.sink.KuduWriter$$anonfun$3.apply(KuduWriter.scala:53)
        at com.datamountaineer.streamreactor.connect.kudu.sink.KuduWriter$$anonfun$3.apply(KuduWriter.scala:53)
        at scala.util.Try$.apply(Try.scala:192)
        at com.datamountaineer.streamreactor.connect.kudu.sink.KuduWriter.<init>(KuduWriter.scala:53)
        at com.datamountaineer.streamreactor.connect.kudu.sink.KuduWriter$.apply(KuduWriter.scala:45)
        at com.datamountaineer.streamreactor.connect.kudu.sink.KuduSinkTask.start(KuduSinkTask.scala:58)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.initializeAndStart(WorkerSinkTask.java:246)
        at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:147)
        at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:146)
        at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:190)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassNotFoundException: io.confluent.kafka.schemaregistry.client.rest.RestService
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at org.apache.kafka.connect.runtime.isolation.PluginClassLoader.loadClass(PluginClassLoader.java:62)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        ... 17 more
[2018-03-30 17:24:40,514] **ERROR Task is being killed and will not recover until manually restarted** (org.apache.kafka.connect.runtime.WorkerTask:149)
[2018-03-30 17:24:40,515] **INFO Stopping Kudu sink.** (com.datamountaineer.streamreactor.connect.kudu.sink.KuduSinkTask:78)

		
So it can see the kudu table, but then stops: why is it looking for the Schema Registry, if no Avro and Autocreate is requested?

Thanks a lot,

Paolo

**Comment by andrewstevenson** [comment]
@pmid What distribution of Kafka are you using? Its missing a the client lib for Confluents Schema registry rather than the code itself.

**Comment by pmid** [comment]
Hi @andrewstevenson, it’s the Cloudera distribution (equivalent to version 0.11.0), so it’s almost surely missing the libs related to schema registry.
I’ll try to retrieve the registry client jar, as you suggest.
I will be able to verify the outcome only on next tuesday, but I am sure it will fix the problem.

Thanks a lot!

**Comment by andrewstevenson** [comment]
@pmid let us know how you get on

**Comment by pmid** [comment]
Hi @andrewstevenson , 

I added the kafka-schema-registry-client-4.0.0.jar, as you suggested, and this fixed the issue I had.

Then I faced a new error: 

"ERROR Encountered error class java.util.HashMap is not valid. Expecting a Struct (com.datamountaineer.streamreactor.connect.kudu.sink.KuduWriter:62)
java.lang.RuntimeException: class java.util.HashMap is not valid. Expecting a Struct"

but it was a matter of configuring properly the key and value converters, as suggested in 

[JMS Sink - Error when using schemaless JSON](https://github.com/Landoop/stream-reactor/issues/137)

[kafka connect hdfs sink connector is failing even when json data contains schema and payload field
](https://stackoverflow.com/questions/43976874/kafka-connect-hdfs-sink-connector-is-failing-even-when-json-data-contains-schema)

I think the issue can be closed - thanks a lot for your support.


Paolo
